src/main/java/io/apimatic/core/security/HmacSignatureVerifier.java

```diff
@@ -0,0 +1,217 @@
+                        .filter(e -> e.getKey() != null
+                                && e.getKey().equalsIgnoreCase(signatureHeaderName))
+                        .map(Map.Entry::getValue)
+                        .findFirst()
+                        .orElse(null);
+
+                if (headerValue == null) {
+                    return VerificationResult.failure(
+                            "Signature header '" + signatureHeaderName + "' is missing.");
+                }
+
+                byte[] provided = extractSignature(headerValue);
+                if (provided == null) {
+                    return VerificationResult.failure(
+                            "Malformed signature header '" + signatureHeaderName + "'.");
+                }
+
+                byte[] message = requestBytesResolver.apply(request);
+                // HMAC per call (thread-safe)
+                Mac mac = Mac.getInstance(algorithm);
+                mac.init(keySpec);
+                byte[] computed = mac.doFinal(message);
+
+                return MessageDigest.isEqual(provided, computed)
+                        ? VerificationResult.success()
+                        : VerificationResult.failure("Signature verification failed.");
+            } catch (Exception ex) {
+                return VerificationResult.failure("Exception: " + ex.getMessage());
+            }
+        });
+    }
+
+    /**
+     * Extracts the digest value from the signature header according to the template
+     * and decodes the signature from the header value.
+     *
+     * @param headerValue The value of the signature header.
+     * @return The decoded signature as a byte array, or null if extraction fails.
+     */
+    private byte[] extractSignature(final String headerValue) {
+        try {
+            int index = signatureValueTemplate.indexOf(SIGNATURE_VALUE_PLACEHOLDER);
+            if (index < 0) {
+                return null;
+            }
+
+            String prefix = signatureValueTemplate.substring(0, index);
+            String suffix = signatureValueTemplate.substring(
+                    index + SIGNATURE_VALUE_PLACEHOLDER.length());
+
+            // find prefix anywhere (case-insensitive)
+            int prefixAt = indexOfIgnoreCase(headerValue, prefix, 0);
+            if (prefixAt < 0) {
+                return null;
+            }
+
+            int digestStart = prefixAt + prefix.length();
+
+            // find suffix after the digest start (case-insensitive)
+            final int digestEnd;
+            if (suffix.isEmpty()) {
+                digestEnd = headerValue.length();
+            } else {
+                digestEnd = indexOfIgnoreCase(headerValue, suffix, digestStart);
+                if (digestEnd < 0) {
+                    return null;
+                }
+            }
+
+            if (digestEnd < digestStart) {
+                return null;
```

Comment:
We aren't covering following branch in tests

Reply:
This code can't be reached